include/zephyr/drivers/intc.h

```diff
@@ -0,0 +1,462 @@
+};
+
+/** @endcond */
+
+/** Test if IRQ specification exists */
+static inline bool intc_irq_spec_exists(const struct intc_irq_spec *spec)
+{
+	return spec->intc != NULL;
+}
+
+/**
+ * @brief Request interrupt request
+ *
+ * @param irq Interrupt request handle
+ * @param spec Interrupt request specification
+ * @param handler Interrupt request handler
+ *
+ * @note The irq handle and the irq spec must remain valid until released
+ *
+ * @retval 0 if successful
+ * @retval -errno code if failure
+ */
+static inline int intc_request_irq(struct intc_irq *irq, const struct intc_irq_spec *spec,
+				   intc_irq_handler_t handler)
+{
+	const struct intc_driver_api *api = (const struct intc_driver_api *)spec->intc->api;
+
+	irq->handler = handler;
+	irq->spec = spec;
+	return api->request_irq(irq);
+}
+
+/**
+ * @brief Enable interrupt request
+ *
+ * @param spec Interrupt request specification
+ *
+ * @retval 0 if successful
+ * @retval -errno code if failure
+ */
+static inline int intc_enable_irq(const struct intc_irq_spec *spec)
+{
+	const struct intc_driver_api *api =
+		(const struct intc_driver_api *)spec->intc->api;
+
+	return api->enable_irq(spec);
+}
+
+/**
+ * @brief Disable interrupt request
+ *
+ * @param spec Interrupt request specification
+ *
+ * @retval 0 if successful
+ * @retval -errno code if failure
+ */
+static inline int intc_disable_irq(const struct intc_irq_spec *spec)
+{
+	const struct intc_driver_api *api =
+		(const struct intc_driver_api *)spec->intc->api;
+
+	return api->disable_irq(spec);
+}
+
+/**
+ * @brief Test if interrupt request is enabled
+ *
+ * @param spec Interrupt request specification
+ *
+ * @retval 1 if interrupt request is enabled
+ * @retval 0 if interrupt request is disabled
+ * @retval -errno code if failure
+ */
+static inline int intc_irq_is_enabled(const struct intc_irq_spec *spec)
+{
+	const struct intc_driver_api *api =
+		(const struct intc_driver_api *)spec->intc->api;
+
+	return api->irq_enabled(spec);
+}
+
+/**
+ * @brief Trigger interrupt request
+ *
+ * @param spec Interrupt request specification
+ *
+ * @retval 0 if successful
+ * @retval -errno code if failure
+ */
+static inline int intc_trigger_irq(const struct intc_irq_spec *spec)
+{
+	const struct intc_driver_api *api =
+		(const struct intc_driver_api *)spec->intc->api;
+
+	return api->trigger_irq(spec);
+}
+
+/**
+ * @brief Release interrupt request
+ *
+ * @param irq Interrupt request handle
+ *
+ * @retval 0 if successful
+ * @retval -errno code if failure
+ */
+static inline int intc_release_irq(struct intc_irq *irq)
+{
+	const struct intc_driver_api *api =
+		(const struct intc_driver_api *)irq->spec->intc->api;
+
+	return api->release_irq(irq);
+}
+
+/**
+ * @brief Get the name of an interrupt request handler
+ *
+ * @param ord Interrupt controller's devicetree ordinal
+ * @param irq Interrupt line number
+ *
+ * Expands to __intc_irq_handler_<ord>_<irq>
+ */
+#define INTC_IRQ_HANDLER_NAME(ord, irq) \
```

Comment:
Is this meant to be used anywhere else than in the macro below? I am asking because it requires very strict values here and it seems little risky to reveal the macro as an part of the API.

Reply:
Its meant to be used within device drivers to populate their interrupt vector tables, maybe it can be moved to a less user facing header like direct_irq_handler.h :)